pkg/controller/jobs/job/job_controller_test.go

```diff
@@ -1804,6 +1805,101 @@ func TestReconciler(t *testing.T) {
 					Obj(),
 			},
 		},
+		"the workload is not admitted and tolerations change": {
+			job: *baseJobWrapper.Clone().Toleration(corev1.Toleration{
+				Key:      "tolarationkey2",
+				Operator: corev1.TolerationOpExists,
+				Effect:   corev1.TaintEffectNoSchedule,
+			}).
+				Obj(),
+			wantJob: *baseJobWrapper.Clone().Toleration(corev1.Toleration{
+				Key:      "tolarationkey2",
+				Operator: corev1.TolerationOpExists,
+				Effect:   corev1.TaintEffectNoSchedule,
+			}).Obj(),
+			workloads: []kueue.Workload{
+				*utiltesting.MakeWorkload(GetWorkloadNameForJob(baseJobWrapper.Name), "ns").
+					Finalizers(kueue.ResourceInUseFinalizerName).
+					Queue("foo").
+					PodSets(
+						*utiltesting.MakePodSet("main", 10).
+							Toleration(corev1.Toleration{
+								Key:      "tolarationkey1",
+								Operator: corev1.TolerationOpExists,
+								Effect:   corev1.TaintEffectNoSchedule,
+							}).
+							Request(corev1.ResourceCPU, "1").
+							Obj(),
+					).
+					Labels(map[string]string{
+						controllerconsts.JobUIDLabel: "",
+					}).
+					Priority(0).
+					Obj(),
+			},
+			wantWorkloads: []kueue.Workload{
+				*utiltesting.MakeWorkload(GetWorkloadNameForJob(baseJobWrapper.Name), "ns").
+					Finalizers(kueue.ResourceInUseFinalizerName).
+					Queue("foo").
+					PodSets(
+						*utiltesting.MakePodSet("main", 10).
+							Toleration(corev1.Toleration{
+								Key:      "tolarationkey2",
+								Operator: corev1.TolerationOpExists,
+								Effect:   corev1.TaintEffectNoSchedule,
+							}).
+							Request(corev1.ResourceCPU, "1").
+							Obj(),
+					).
+					Labels(map[string]string{
+						controllerconsts.JobUIDLabel: "",
+					}).
+					Priority(0).
+					Obj(),
+			},
+		},
+		"the workload is admitted and tolerations change": {
+			job: *baseJobWrapper.Clone().Toleration(corev1.Toleration{
+				Key:      "tolarationkey2",
+				Operator: corev1.TolerationOpExists,
+				Effect:   corev1.TaintEffectNoSchedule,
+			}).
+				Suspend(false).
+				Obj(),
+			wantJob: *baseJobWrapper.Clone().Toleration(corev1.Toleration{
+				Key:      "tolarationkey1",
```

Comment:
So we are undoing the user changes. Not ideal, but somewhat fair. Can you make sure that there is an event for the suspension, indicating that the queued workload doesn't match the job? This way the user knows that Kueue took some action and they can fix whatever they wanted to do.

Reply:
My biggest concern around this is that if the scheduler is fast and readmits the workload the user will never be able to change the toleration. 
 Maybe we can / should add a configuration key to skip restoring from an unmatching workload.

Reply:
The source of the restore info is mentioned in the stop event message now.